### Introduction to Confidence Intervals

#### Theoretical Background:
In the previous lesson, we introduced inferential statistics, and the concept of sampling distributions, along with one of the most important theorems in statistics, that is the **Central Limit Theorem**.
In this lesson we will explore and make you familiar with **Confidence intervals**.

<br> 

Confidence intervals are a range of values, derived from sample data, that are used to estimate the value of an unknown population parameter. Unlike a single point estimate, a confidence interval provides an interval estimate that conveys <b>the degree of uncertainty associated with the parameter estimate</b>. The most common confidence level for intervals is 95%, but other levels (such as 90% or 99%) can also be used depending on the context.

The width of the confidence interval gives us an idea of how precise our estimate is. Narrower intervals represent more precise estimates. The confidence level tells us how confident we can be that the interval includes the true parameter value. For example, a 95% confidence interval means that if we were to take 100 different samples and compute a 95% confidence interval for each sample, we would expect about 95 of those intervals to contain the population parameter.

![](https://github.com/IranelGV/Ironhack_IDGV/blob/main/Confidential_Intervals/Confidential_Intervals.jpeg?raw=true)



A basic formula for constructing a confidence interval for a population mean, when the population standard deviation is known, is given by:

$$
\text{Confidence Interval} = \bar{x} \pm Z \cdot \left( \frac{\sigma}{\sqrt{n}} \right)
$$


where:
- $\bar{x}$ is the sample mean,
- $z$ is the Z-score associated with the desired confidence level. Put simply, a Z-score is a way of measuring how many standard deviations an element is from the mean,
- $\sigma$ is the population standard deviation, and
- $n$ is the sample size.
- $\frac{\sigma}{\sqrt{n}}$ is known as the standard error of the mean. It tells us on average, how big of a mistake we will make when trying to estimate the population mean with the average of a sample

So essentially, the “confidence level” is how sure we are that the range we created includes the **true value of the parameter**. In the case of a 95% confidence interval, say for the mean weight of squirrels, as 1.1kg to 1.3kg, we're saying we're 95% confident that the true mean weight of all squirrels lies somewhere between 1.1 and 1.3 kilograms.

When the population standard deviation is unknown and the sample size is small, the t-distribution is used instead of the normal distribution, and the sample standard deviation is used in place of the population standard deviation.

### Practical example

Now let's create a Python example that calculates a 95% confidence interval for the average weight of a sample of apples. Assume we have weighed a sample of apples from a large orchard, and we want to estimate the average weight of apples in the entire orchard based on our sample.

Sample Data:
Suppose we have a sample of 50 apples.
The average (mean) weight of the apples in our sample is 180 grams.
The standard deviation of the weights in our sample is 20 grams.
Objective:
Calculate the 95% confidence interval for the average weight of apples in the orchard.

**In simpler terms, we want to create an interval that will include, with a 95% confidence level, the true mean weight of the apples in the whole orchard!**

**In the cell below, the code is ready for you to run:**

In [1]:
#Import libraries
import numpy as np
import scipy.stats as stats

# Here are our sample statistics
sample_mean = 180  # The mean weight of the sampled apples in grams
sample_std_dev = 20  # The standard deviation of the sample in grams
sample_size = 50 # The sample size
confidence_level = 0.95 # Our 95% confidence level

# We calculate the standard error of the mean
standard_error = sample_std_dev / np.sqrt(sample_size)

# We calculate the margin of error using the t-distribution
# Since the population standard deviation is unknown, we use the t-distribution
t_critical = stats.t.ppf((1 + confidence_level) / 2, df=sample_size-1)  # t-critical value for 95% confidence

margin_of_error = t_critical * standard_error

# Given our sample mean and margin of error, we calculate the confidence interval!
confidence_interval = (sample_mean - margin_of_error, sample_mean + margin_of_error)

print(f"The 95% confidence interval for the average weight of apples in the orchard is: {confidence_interval[0]:.2f} to {confidence_interval[1]:.2f} grams")


The 95% confidence interval for the average weight of apples in the orchard is: 174.32 to 185.68 grams


**Great! We managed to create a 95% confidence interval for the mean weight of the apples in the entire orchard, based on one sample of 100 apples!**
**We can say, with 95% of certainty, that it lies somewhere between 174.32 and 185.68 grams!**

**Your turn now! Suppose you are working for Nike, more specifically for their Air Jordan department in Berlin. 
Create a 95% confidence interval for the shoe size of Berliners given the sample of 100 Berlin shoe sizes you collected.**

<br>
<b>Here are the statistics from the sample you have</b>:

<br>
sample_mean = 41 
<br>
sample_std_dev = 3 
<br>
sample_size = 100
<br>
confidence_level = 0.95

**Feel free to use the code provided in the example above, you will just have to tweak the sample statistics**

In [3]:
#Your code goes here

#Import libraries
import numpy as np
import scipy.stats as stats

# Here are our sample statistics
sample_mean = 41  # The mean weight of the sampled apples in grams
sample_std_dev = 3  # The standard deviation of the sample in grams
sample_size = 100 # The sample size
confidence_level = 0.95 # Our 95% confidence level

# We calculate the standard error of the mean
standard_error = sample_std_dev / np.sqrt(sample_size)

# We calculate the margin of error using the t-distribution
# Since the population standard deviation is unknown, we use the t-distribution
t_critical = stats.t.ppf((1 + confidence_level) / 2, df=sample_size-1)  # t-critical value for 95% confidence

margin_of_error = t_critical * standard_error

# Given our sample mean and margin of error, we calculate the confidence interval!
confidence_interval = (sample_mean - margin_of_error, sample_mean + margin_of_error)

print(f"The 95% confidence interval for the average weight of apples in the orchard is: {confidence_interval[0]:.2f} to {confidence_interval[1]:.2f} grams")





The 95% confidence interval for the average weight of apples in the orchard is: 40.40 to 41.60 grams


Here are some useful links to understand confidence intervals better:
    <br> - https://www.investopedia.com/terms/c/confidenceinterval.asp
    <br> - https://www.youtube.com/watch?v=TqOeMYtOc1w